File: Pairs-with-K-difference.java
import java.util.HashMap;
import java.util.Map;

// Time Complexity : O(n) where n is # elements in nums
// Space Complexity : O(n) where n is # elements in nums to be stored in map
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int findPairs(int[] nums, int k) {
        
        // can't have abs diff less than 0 or arr empty
        if (k < 0 || nums.length == 0) {
            return 0;
        }
        
        // we store elements in num as KEY
        // and their count as VALUE
        Map<Integer, Integer> map = new HashMap<>();
        
        int count = 0;
        
        for (int num : nums) {
            map.put (num, map.getOrDefault(num, 0) + 1);
        }
        
        for (int i : map.keySet()) {
            
            // here we check the number of duplicate elements in nums
            if (k == 0) {
                if (map.get(i) >= 2) {
                    count++;
                }
            }
            
            else {
                if (map.containsKey(i + k)) {
                    count++;
                }
            }
        }
        
        return count;
    }
}

File: pairs_with_k_difference.py
# TC - O(n)
# SC - O(n)

class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        if k < 0:
            return 0
        
        freq_map = {}
        for num in nums:
            freq_map[num] = freq_map.get(num, 0) + 1
        
        count = 0
        if k == 0:
            for freq in freq_map.values():
                if freq > 1:
                    count += 1
        else:
            for num in freq_map.keys():
                if num + k in freq_map:
                    count += 1
        
        return count
